File: 3615.longest-palindromic-path-in-graph.py
#
# @lc app=leetcode id=3615 lang=python3
#
# [3615] Longest Palindromic Path in Graph
#
# @lc code=start
class Solution:
    def maxLen(self, n: int, edges: List[List[int]], label: str) -> int:
        # Build adjacency list
        graph = [[] for _ in range(n)]
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)
        
        max_length = 1  # At least one node
        
        def dfs(node, visited_mask, path):
            nonlocal max_length
            
            # Check if current path is palindrome
            if path == path[::-1]:
                max_length = max(max_length, len(path))
            
            # Try to extend the path
            for neighbor in graph[node]:
                if not (visited_mask & (1 << neighbor)):
                    dfs(neighbor, visited_mask | (1 << neighbor), path + label[neighbor])
        
        # Try starting from each node
        for start in range(n):
            dfs(start, 1 << start, label[start])
        
        return max_length
# @lc code=end

File: 3615.longest-palindromic-path-in-graph.cpp
#
# @lc app=leetcode id=3615 lang=cpp
#
# [3615] Longest Palindromic Path in Graph
#
# @lc code=start
class Solution {
public:
    int maxLen(int n, vector<vector<int>>& edges, string label) {
        vector<vector<int>> adj(n);
        for (auto& e : edges) {
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }
        
        vector<vector<pair<int, int>>> states(1 << n);
        vector<vector<vector<bool>>> dp(1 << n, vector<vector<bool>>(n, vector<bool>(n, false)));
        
        auto addState = [&](int mask, int i, int j) {
            if (!dp[mask][i][j]) {
                dp[mask][i][j] = true;
                states[mask].push_back({i, j});
            }
        };
        
        int ans = 1;
        
        // Base case: single nodes (palindrome of length 1)
        for (int i = 0; i < n; i++) {
            addState(1 << i, i, i);
        }
        
        // Base case: two adjacent nodes with same label (palindrome of length 2)
        for (int i = 0; i < n; i++) {
            for (int j : adj[i]) {
                if (label[i] == label[j]) {
                    addState((1 << i) | (1 << j), i, j);
                    ans = 2;
                }
            }
        }
        
        // Extend palindromes by adding matching pairs at both ends
        for (int mask = 1; mask < (1 << n); mask++) {
            for (auto& [i, j] : states[mask]) {
                for (int u : adj[i]) {
                    if (mask & (1 << u)) continue;
                    for (int v : adj[j]) {
                        if (mask & (1 << v)) continue;
                        if (u == v) continue;
                        if (label[u] != label[v]) continue;
                        
                        int newMask = mask | (1 << u) | (1 << v);
                        addState(newMask, u, v);
                        ans = max(ans, __builtin_popcount(newMask));
                    }
                }
            }
        }
        
        return ans;
    }
};
# @lc code=end